File: primalTestDotNet/Entities/Monster.cs
using PrimalTestDotNet.Level;

namespace PrimalTestDotNet.Entities;

class Monster(IGameObjectContainer container, int x, int y) : IGameObject
{
    public char Sprite => 's';

    public IntVector2 Position { get; } = new(x, y);

    public void Draw(EntityRenderer renderer)
    {
        renderer.Draw(this);
    }

    public void Interact(Hero hero)
    {
        if (hero.HasSword)
        {
            hero.Health--;
            container.RemoveGameObject(this);
        }
        else
        {
            hero.Kill();
        }
    }
}

File: PrimalTestCpp/Monster.cpp
#include "Monster.h"
#include "EntityRenderer.h"
#include "Hero.h"
#include "Level.h"

Monster::Monster(Level& level, int x, int y) : _level(level), _position(x, y) { }

char Monster::getSprite() const
{
	return 's';
}

const IntVector2D& Monster::getPosition() const
{
	return _position;
}

void Monster::draw(const EntityRenderer& entityRenderer) const
{
	entityRenderer.draw(*this);
}

void Monster::interact(Hero& hero)
{
	if (hero.hasSword()) {
		hero.setHealth(hero.getHealth() - 1);
		_level.removeGameObject(*this);
	}
	else {
		hero.kill();
	}
}
